src/main/java/com/fasterxml/jackson/core/json/ByteSourceJsonBootstrapper.java

```diff
@@ -230,6 +232,13 @@ public Reader constructReader() throws IOException
                 InputStream in = _in;
 
                 if (in == null) {
+                    int size = _inputEnd - _inputPtr;
+                    if (size >= 0 && size <= 8192) {
+                        // [jackson-core#488] Avoid overhead of heap ByteBuffer allocated by InputStreamReader
+                        // when processing small inputs up to 8KiB.
+                        Charset charset = Charset.forName(enc.getJavaName());
+                        return new CharBufferReader(charset.decode(ByteBuffer.wrap(_inputBuffer, _inputPtr, _inputEnd)));
+                    }
                     in = new ByteArrayInputStream(_inputBuffer, _inputPtr, _inputEnd);
```

Comment:
Open to thoughts here on threshold. 

This new path decoding from `ByteBuffer` to `CharBuffer` will allocate a `char[]` based on the encoding `Charset`, so I chose `8192` as that would be the break even point for `InputStreamReader`'s `StreamDecoder`'s `8192` byte heap `ByteBuffer`.